File: ts/mergeSort.ts



type Comparator<T> = (a: T, b: T) => number; 
const defaultCompare = <T>(a: T, b: T): number => (a < b) ? -1 : (a > b) ? 1 : 0;

function merge<T>(left: T[], right: T[], compare: Comparator<T>): T[]{

    const result: T[] = []
    let i = 0;
    let j = 0;

    while(i < left.length && j < right.length){
        
        if(compare(left[i], right[j]) <= 0){
            result.push(left[i]);
            i++;
        }else{
            result.push(right[j]);
            j++;
        }
    }

    while(right.length != j){
        result.push(right[j]);
        j++;
    }

     while(left.length != i){
        result.push(left[i]);
        i++;
    }

    return result;
}

export function mergeSort<T>(arr: T[], compare: Comparator<T> = defaultCompare): T[]{

    // BASE CASE: 0 or 2 elements - already sorted, stop recursion
    if( arr.length <= 1) return arr;

    // DIVIDE: split into two halves at the midpoint
    const mid = Math.floor(arr.length / 2);
    const left = arr.slice(0 , mid);
    const right = arr.slice(mid);

    //Conquer: recursively sort each half, then merge the sorted halfes
    return merge(mergeSort(left, compare),mergeSort(right, compare), compare);

}

function check(name: string, actual : unknown, expected:unknown): void{

    const a = JSON.stringify(actual);
    const e = JSON.stringify(expected);
    const pass = a === e;

    console.log(`${pass ? "Pass" : "Fail"} ${name}`);
    console.log(`   got:      ${a}`);
    if(!pass) console.log(`   expected: ${e}`);

}

check("ordena números", mergeSort([5, 2, 8, 1, 3]), [1, 2, 3, 5, 8]);
check("arreglo vacío", mergeSort([]), []);
check("un solo elemento", mergeSort([7]), [7]);

File: c++/DivideAndConquer/SortAlgorithms/MergeSort/mergeSort.cpp
//
// Created by sebas on 07/09/2026.
//

#include <iostream>
#include <vector>
#include <cmath>

using namespace std;

vector<int> merge(const vector<int>& left , const vector<int>& right) {

    vector<int> result;
    int i =0 , j = 0;

    while (i < left.size() && j < right.size()) {


        if (left[i] <= right[j]) {

            result.push_back(left[i]);
            i++;
        }
        else {
            result.push_back(right[j]);
            j++;

        }
    }

    while (i < left.size()) {
        result.push_back(left[i]);
        i++;
    }

    while (j < right.size()) {
        result.push_back(right[j]);
        j++;
    }
    return result;
}

vector<int> mergeSort(const vector<int>& arr) {

    if (arr.size() <= 1) return arr;
    int mid = floor(arr.size()/2);
    vector<int> left (arr.begin(), arr.begin() + mid);
    vector<int> right (arr.begin() + mid, arr.end());

    return merge(mergeSort(left), mergeSort(right));
}

vector<int> nums = {2,4,6,1,3,5,7,8,0,9,10};

int main() {

    cout<<"Lista no ordenada= ";
    for (int x : nums) cout <<x << " ";

    vector<int> result = mergeSort(nums);

    cout<<"\nLista ordenada= ";
    for (int x : result) cout <<x<<" ";

    return 0;
}